### PR TITLE
Support for the meson build system

### PR DESCRIPTION
**Pull request by dagle**
This add support for the meson build system which supports multiple operating systems among things out of the box. This requires https://github.com/jstedfast/gmime/pull/141 to be merged, because gnome.generate_gir() treats warnings as errors.

### CODE REVIEW

**Comment by dagle** [comment]
This pr does not contain building documentation, which will be included in a future pr. My reasoning for this is that I updated the documentation to build using the new gi-docgen instead of gtk-doc.

coverage works but reports false atm since I don't know if there is a portable way to check for coverage packages / support.

profiling support isn't included in this pr.

File: meson.build [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+  conf_data.set('HAVE_ICONV_DETECT_H', 1)
+else
+  # not cross-compiler secure
+  root = include_directories('.')
+  has_detect = cc.run(iconv_detect, name: 'iconv-detect', 
+    include_directories: root).returncode() == 0
+  if has_detect
+    conf_data.set('HAVE_ICONV_DETECT_H', 1)
+  endif
+endif
```
Comment (eli-schwartz) [inline meson.build:81]: Meson can absolutely handle this with cross compilation, you just have to define an exe_wrapper (usually qemu-user). If you don't, meson errors out with "Can not run test applications in this cross environment." which... I suppose could be enhanced to mention exe_wrapper, hmm.

But this is a pretty awkward way to do things here. You're running a test program in the source directory and emitting unclean output into the source directory instead of the build directory -- this violates Meson's guarantee of out-of-source build directories, potentially messing up situations such as read-only mounted source directories or having multiple build directories, one native one cross compiled, which need *different* generated headers.

File: meson.build [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    { 'm': 'HAVE_NETDB_H',            'v': 'netdb.h', },
+    { 'm': 'HAVE_STDINT_H',           'v': 'stdint.h', },
+    { 'm': 'HAVE_WINDOWS_H',          'v': 'windows.h'},
+    { 'm': 'HAVE_SYS_MMAN_H',         'v': 'sys/mman.h', },
+    { 'm': 'HAVE_SYS_PARAM_H',        'v': 'sys/param.h', },
+  ]
+  if cc.has_header(hd['v'])
+    conf_data.set10(hd['m'], true)
+  endif
+endforeach
```
Comment (eli-schwartz) [inline meson.build:95]: This could be a flat list of headers. e.g. if you use Meson's `tools/ac_converter.h`, it suggests the following:

```meson
cc = meson.get_compiler('c')
cdata = configuration_data()
check_headers = [
  'dlfcn.h',
  'getopt.h',
  'iconv/detect.h',
  'inttypes.h',
  'netdb.h',
  'poll.h',
  'stdint.h',
  'stdio.h',
  'stdlib.h',
  'strings.h',
  'string.h',
  'sys/mman.h',
  'sys/param.h',
  'sys/stat.h',
  'sys/types.h',
  'time.h',
  'unistd.h',
  'winsock2.h',
]

foreach h : check_headers
  if cc.has_header(h)
    cdata.set('HAVE_' + h.underscorify().to_upper(), 1)
  endif
endforeach

check_functions = [
  ['HAVE_FSYNC', 'fsync', '#include<unistd.h>'],
  ['HAVE_GETADDRINFO', 'getaddrinfo', '#include<netdb.h>'],
# check token ['HAVE_GETDOMAINNAME']
# check token ['HAVE_GETHOSTNAME']
  ['HAVE_GETPAGESIZE', 'getpagesize', '#include<unistd.h>'],
  ['HAVE_MMAP', 'mmap', '#include<sys/mman.h>'],
# check token ['HAVE_MSYNC']
# check token ['HAVE_MUNMAP']
  ['HAVE_POLL', 'poll', '#include<poll.h>'],
  ['HAVE_SELECT', 'select', '#include<sys/select.h>'],
# check token ['HAVE_UTSNAME_DOMAINNAME']
]

foreach f : check_functions
  if cc.has_function(f.get(1), prefix : f.get(2))
    cdata.set(f.get(0), 1)
  endif
endforeach

cdata.set('SIZEOF_OFF_T', cc.sizeof('off t'))
cdata.set('SIZEOF_SIZE_T', cc.sizeof('size_t'))
cdata.set('SIZEOF_SSIZE_T', cc.sizeof('ssize_t'))
cdata.set('SIZEOF_TIME_T', cc.sizeof('time t'))

configure_file(input : 'config.h.meson',
  output : 'config.h',
  configuration : cdata)
```

Note the use of:
```meson
cdata.set('HAVE_' + h.underscorify().to_upper(), 1)
```

File: gmime/meson.build [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+
+pkg = import('pkgconfig')
+
+pkg.generate(libgmime,
+  name: meson.project_name(),
+  filebase: gmime_package,
+  subdirs: gmime_package,
+  description: 'MIME parser and utility library',
+  requires: LIBS,
+  # libraries_private: EXTRA_LIBS,
```
Comment (eli-schwartz) [inline gmime/meson.build:185]: LIBS is a list of strings. Consider using the returned `dependency()` objects as those automatically do "the right thing" (including enforce the same version constraints you used when looking up the dependency to begin with).

File: gmime/meson.build [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+version_data.set('GMIME_BINARY_AGE', GMIME_BINARY_AGE)
+version_data.set('GMIME_INTERFACE_AGE', GMIME_INTERFACE_AGE)
+configure_file(input: 'gmime-version_meson.h.in',
+  output: 'gmime-version.h',
+  configuration : version_data
+)
+
+gen_table = executable('gen-table', 'gen-table.c')
+charset_map = executable('charset-map', 'charset-map.c', dependencies : deps, 
+  include_directories: inc_conf)
```
Comment (eli-schwartz) [inline gmime/meson.build:158]: Since cross compilation came up... you can mark an executable as `native: true` to make it be compiled with the build machine compiler instead of the cross-compiler, although if you need an exe_wrapper anyway that might not matter so much.

File: meson.build [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+GMIME_MINOR_VERSION=ver_arr[1].to_int()
+GMIME_MICRO_VERSION=ver_arr[2].to_int()
+GMIME_BINARY_AGE= 100 * GMIME_MINOR_VERSION + GMIME_MICRO_VERSION
+GMIME_INTERFACE_AGE=0
+
+gmime_package = 'gmime-@0@.0'.format(GMIME_MAJOR_VERSION)
+
+add_global_arguments('-DHAVE_CONFIG_H=1', language : 'c')
+
+iconv = dependency('iconv', required : false, method: get_option('iconv'))
```
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply meson.build:22]: By the way, marking it as `required: false` seems wrong. The configure.ac considers it mandatory, and the source code unconditionally tries to use it. All you're doing here is adding it to the list of deps -- you don't check whether it's available and set different preprocessor defines to e.g. avoid using it if it doesn't exist.

File: meson.build [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+  glib,
+  gpgme,
+  gobject,
+  dependency('gio-2.0'),
+  libidn2,
+  iconv,
+]
+
+gmime_prefix = get_option('prefix')
+gmime_datadir = join_paths(gmime_prefix, get_option('datadir'))
```
Comment (eli-schwartz) [inline meson.build:38]: This doesn't appear to ever be used, nor can I find anything in the source code or autotools that it would correspond to.

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: meson.build (hunk 5 above)
[the remark this answers is not in the document]
Comment (dagle) [reply meson.build]: Yeah, required: false is wrong. It was something left over. Removed it.

The reason for the get_option is to be able to specify what version of iconv to use if you have multiple installed. This is something configure.ac included. Maybe it's considered bad practice or there is a better way to do it in now?

File: meson.build [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+conf_data.set_quoted('GMIME_VERSION', meson.project_version())
+
+LIBS = ['glib-2.0', 'gio-2.0', 'gobject-2.0' ]
+EXTRA_LIBS = ['zlib']
+
+if cc.has_header('getopt.h')
+  if cc.has_function('getopt_long')
+    conf_data.set('HAVE_GETOPT_H', 1)
+  endif
+endif
```
[the remark this answers is not in the document]
Comment (dagle) [reply meson.build]: True. Fixed

File: meson.build (hunk 1 above)
Comment (dagle) [reply meson.build]: Yeah. Adding an error if we can't run. 

In the long run I think it might be better to not pollute the system. The reason for me to "port the awkwardness" over is to not disrupt the autotools build which can require this file to exist.

File: meson.build (hunk 2 above)
Comment (dagle) [reply meson.build]: Cool, never heard of that tool. Fixed

File: gmime/meson.build (hunk 3 above)
Comment (dagle) [reply gmime/meson.build:185]: Fixed

File: gmime/meson.build (hunk 4 above)
Comment (dagle) [reply gmime/meson.build]: fixed

File: meson.build (hunk 6 above)
Comment (dagle) [reply meson.build]: Ah. Those are used for the docs, which are not included in this PR. Removing.

File: meson.build (hunk 5 above)
[the remark this answers is not in the document]
Comment (eli-schwartz) [reply meson.build]: I have no idea what the original intention was, but the general rule of thumb is that you can either get the standalone gnu one (always an external library) or rely on the platform one (which may be a builtin or an external library, and for glibc it's also the same underlying codebase as the gnu one).

Which one you use isn't much of an option, though, because they both require the same `iconv.h` header and if you have both available then the gnu one overrides the header search path. It also overrides the symbol names, so if you don't perform a link verification test using the header (the dependency() provided by meson does this for you) you may end up compiling against the default header which is the gnu one, but linking to an incompatible BSD or musl one that is built into your libc.

The only choice you have really is to either:
- use the one picked up by auto
- return an error or simply disable iconv support if you didn't get the one you wanted
- modify CFLAGS and LDFLAGS *before* running meson, to prioritize a custom copy in the location you prefer


You're effectively offering an opt-in for option 2 (return a fatal error). But I traditionally wonder: does it actually matter which iconv you find?

File: meson.build (hunk 1 above)
Comment (rossburton) [reply meson.build]: As someone who cross-compiles daily, remember to include a fallback for cross and can't run by checking for a property in the cross file.

Comment (eli-schwartz) [reply meson.build]: To be fair, this code only runs in the else branch of a `get_option()`...

File: meson.build (hunk 5 above)
[the remark this answers is not in the document]
Comment (jstedfast) [reply meson.build]: > I have no idea what the original intention was, but the general rule of thumb is that you can either get the standalone gnu one (always an external library) or rely on the platform one (which may be a builtin or an external library, and for glibc it's also the same underlying codebase as the gnu one).

This was the original intention, yes.

**Comment by jstedfast** [comment]
First, sorry for taking so long to get around to this, I'm just so busy these days and have multiple projects that I maintain in my limited free time.

I also have no knowledge of the meson build system or what GNOME is doing these days.

That said, I just had the most awful experience trying to build gmime on macOS (the closest thing I have to a suitable for gmime development environment) which is what made me think of this PR that I'd been meaning to get around to.

I just installed meson from `brew` on my macOS machine and cloned this branch.

How do I test this?

**Comment by eli-schwartz** [comment]
tl;dr 

```bash
# replaces ./configure
meson setup builddir/

# replaces /configure --enable-foo
meson setup builddir/ -Dfoo=enabled

# replaces ./configure --help
meson configure

# replaces make
ninja
```

**Comment by jstedfast** [comment]
```
[jestedfa@localhost gmime]$ meson setup builddir && cd builddir
The Meson build system
Version: 1.4.1
Source dir: /Users/jestedfa/src/gmime
Build dir: /Users/jestedfa/src/gmime/builddir
Build type: native build
Project name: GMime
Project version: 3.2.13
C compiler for the host machine: cc (clang 15.0.0 "Apple clang version 15.0.0 (clang-1500.3.9.4)")
C linker for the host machine: cc ld64 1053.12
Host machine cpu family: aarch64
Host machine cpu: aarch64
Run-time dependency iconv found: YES
Found pkg-config: YES (/opt/homebrew/bin/pkg-config) 0.29.2
Run-time dependency libidn2 found: YES 2.3.7
Run-time dependency gpgme found: YES 1.23.2
Run-time dependency glib-2.0 found: YES 2.80.3
Run-time dependency gobject-2.0 found: YES 2.80.3
Run-time dependency gio-2.0 found: YES 2.80.3
Run-time dependency zlib found: YES 1.2.11
Header "getopt.h" has symbol "getopt_long" : YES 
Checking if "iconv-detect" runs: DID NOT COMPILE
Has header "dlfcn.h" : YES 
Has header "getopt.h" : YES 
Has header "inttypes.h" : YES 
Has header "netdb.h" : YES 
Has header "poll.h" : YES 
Has header "stdint.h" : YES 
Has header "stdio.h" : YES 
Has header "stdlib.h" : YES 
Has header "strings.h" : YES 
Has header "string.h" : YES 
Has header "sys/mman.h" : YES 
Has header "sys/param.h" : YES 
Has header "sys/stat.h" : YES 
Has header "sys/types.h" : YES 
Has header "time.h" : YES 
Has header "unistd.h" : YES 
Has header "winsock2.h" : NO 
Checking whether type "struct utsname" has member "domainname" : NO 
Checking for function "fsync" : YES 
Checking for function "getaddrinfo" : YES 
Checking for function "getdomainname" : YES 
Checking for function "gethostname" : YES 
Checking for function "getpagesize" : YES 
Checking for function "mmap" : YES 
Checking for function "msync" : YES 
Checking for function "munmap" : YES 
Checking for function "poll" : YES 
Checking for function "select" : YES 
Configuring config.h using configuration
Configuring gmime-version.h using configuration

gmime/meson.build:188:0: ERROR: File gmime-version.h does not exist.

A full log can be found at /Users/jestedfa/src/gmime/builddir/meson-logs/meson-log.txt
```

**Comment by jstedfast** [comment]
Looks like the meson setup logic assumes gmime-version.h exists from the onset, but it is/was autogenerated by the old `configure` script.

Presumably meson needs to auto-generate it as well.

**Comment by jstedfast** [comment]
Hmmm, looks like it should be doing that...

```
configure_file(input: 'gmime-version_meson.h.in',
  output: 'gmime-version.h',
  configuration : version_data
)
```

**Comment by jstedfast** [comment]
Okay, I'm guessing that it's because libgmime_headers contains gmime-version.h

I wonder if maybe inc_conf needs to include the meson build directory so that gmime-version.h can be picked up from there?

**Comment by jstedfast** [comment]
Rebased this on top of the latest gmime to fix the gir errors that I was getting when trying to `meson compile`

**Comment by jstedfast** [comment]
`Checking if "iconv-detect" runs: DID NOT COMPILE` is a bit concerning.

Logs:

```
Code:
 #include <iconv.h>

int main() {
    iconv_open("","");
}
-----------
Command line: `cc /Users/jestedfa/src/gmime/builddir/meson-private/tmp4tsmc6zs/testfile.c -o /Users/jestedfa/src/gmime/builddir/meson-private/tmp4tsmc6zs/output.exe -O0 -Werror=implicit-function-declaration` -> 1
stderr:
Undefined symbols for architecture arm64:
  "_iconv_open", referenced from:
      _main in testfile-4a9b3c.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
-----------
Running compile:
Working directory:  /Users/jestedfa/src/gmime/builddir/meson-private/tmphax2pj_h
Code:
 
        #ifdef __has_include
         #if !__has_include("iconv.h")
          #error "Header 'iconv.h' could not be found"
         #endif
        #else
         #include <iconv.h>
        #endif
-----------
Command line: `cc /Users/jestedfa/src/gmime/builddir/meson-private/tmphax2pj_h/testfile.c -E -P -P -O0 -Werror=implicit-function-declaration` -> 0
Running compile:
Working directory:  /Users/jestedfa/src/gmime/builddir/meson-private/tmplrfdf0sy
Code:
 int main(void) { return 0; }

-----------
Command line: `cc /Users/jestedfa/src/gmime/builddir/meson-private/tmplrfdf0sy/testfile.c -o /Users/jestedfa/src/gmime/builddir/meson-private/tmplrfdf0sy/output.exe -O0 -Werror=implicit-function-declaration -liconv -Wl,-undefined,dynamic_lookup` -> 0
Run-time dependency iconv found: YES
```
